ambry-router/src/test/java/com.github.ambry.router/OperationTrackerTest.java

```diff
@@ -923,11 +925,59 @@ public void operationClassTest() {
       // ensure the success target matches the number specified for each type of operaiton
       if (operationTracker != null) {
         assertEquals("The suggest target doesn't match", (long) entry.getValue(),
-            (operationTracker).getSuccessTarget());
+            operationTracker.getSuccessTarget(ReplicaType.DISK_BACKED));
       }
     }
   }
 
+  /**
+   * Test cases with cloud replicas in the local datacenter.
+   */
+  @Test
+  public void cloudReplicaInLocalDcTest() {
+    // test success in cloud dc
+    initializeWithCloudDcs(true);
+    originatingDcName = getDatacenters(ReplicaType.DISK_BACKED).iterator().next();
+    OperationTracker ot =
+        getOperationTracker(true, 2, 3, false, Integer.MAX_VALUE, RouterOperation.GetBlobOperation, false);
+    assertFalse("Operation should not have been done.", ot.isDone());
+    // parallelism of 1 for cloud replicas, so only one request should be sent
+    sendRequests(ot, 1, false);
+    assertFalse("Operation should not have been done.", ot.isDone());
+    ot.onResponse(inflightReplicas.poll(), TrackedRequestFinalState.SUCCESS);
+    // success target of 1 with cloud replicas.
+    assertTrue("Operation should have succeeded", ot.hasSucceeded());
+    assertTrue("Operation should be done", ot.isDone());
+  }
+
+  /**
+   * Test cases with cloud replicas in the remote datacenter.
+   */
+  @Test
+  public void cloudReplicaInRemoteDcTest() {
+    // test failure in disk dc with fallback to cloud DC
+    initializeWithCloudDcs(false);
+    originatingDcName = getDatacenters(ReplicaType.CLOUD_BACKED).iterator().next();
+    OperationTracker ot =
+        getOperationTracker(true, 2, 3, false, Integer.MAX_VALUE, RouterOperation.GetBlobOperation, false);
+    assertFalse("Operation should not have been done.", ot.isDone());
+    // parallelism of 3 for disk replicas (local dc).
+    sendRequests(ot, 3, false);
+    assertFalse("Operation should not have been done.", ot.isDone());
+    ot.onResponse(inflightReplicas.poll(), TrackedRequestFinalState.FAILURE);
+    // one response frees up a spot to send a cloud request
+    sendRequests(ot, 1, false);
+    for (int i = 0; i < 2; i++) {
+      ot.onResponse(inflightReplicas.poll(), TrackedRequestFinalState.FAILURE);
+      // parallelism should have changed to 1 after cloud request was sent
+      sendRequests(ot, 0, false);
+    }
+    assertFalse("Operation should not have been done.", ot.isDone());
+    ot.onResponse(inflightReplicas.poll(), TrackedRequestFinalState.SUCCESS);
+    assertTrue("Operation should have succeeded", ot.hasSucceeded());
+    assertTrue("Operation should be done", ot.isDone());
+  }
+
```

Comment:
Since both of these test cases succeed, shouldn't we also have one that fails?

Reply:
Added some more test cases